go/tao/linux_process_factory.go

```diff
@@ -196,38 +212,137 @@ func (lpf *LinuxProcessFactory) Launch(prog string, args []string, uid, gid int)
+	}
+
+	// Every hosted process is given its own process group (Setpgid=true). This
+	// ensures that hosted processes will not be in orphaned process groups,
+	// allowing them to receive job control signals (SIGTTIN, SIGTTOU, and
+	// SIGTSTP).
+	//
+	// If this host is running in "daemon" mode, i.e. without a controlling tty
+	// and in our own session and process group, then this host will be (a) the
+	// parent of a process in the child's group, (b) in the same session, and
+	// (c) not in the same group as the child, so it will serve as the anchor
+	// that keeps the child process groups from being considered orphaned.
+	//
+	// If this host is running in "foreground" mode, i.e. with a controlling tty
+	// and as part of our parent process's session but in our own process group,
+	// then the same three conditions are satisified, so this host can still
+	// serve as the anchor that keeps the child process groups from being
+	// considered orphaned. (Note: We could also use Setpid=false in this case,
+	// since the host would be part of the child process group and our parent
+	// would then meet the requirements.)
+
 	spa := &syscall.SysProcAttr{
 		Credential: &syscall.Credential{
-			Uid: uint32(uid),
-			Gid: uint32(gid),
+			Uid: uint32(p.spec.Uid),
+			Gid: uint32(p.spec.Uid),
 		},
+		// Setsid: true, // Create session.
+		Setpgid: true, // Set process group ID to new pid (SYSV setpgrp)
+		// Setctty: true, // Set controlling terminal to fd Ctty (only meaningful if Setsid is set)
+		// Noctty: true, // Detach fd 0 from controlling terminal
+		// Ctty: 0, // Controlling TTY fd (Linux only)
 	}
-	cmd := &Process{
-		&exec.Cmd{
-			Path:        prog,
-			Args:        args,
-			Stdin:       os.Stdin,
-			Stdout:      os.Stdout,
-			Stderr:      os.Stderr,
-			Env:         env,
-			ExtraFiles:  extraFiles,
-			SysProcAttr: spa,
-		},
+	p.Cmd = exec.Cmd{
+		Path:        p.Temppath,
+		Dir:         wd,
+		Args:        p.spec.Args,
+		Stdin:       stdin,
+		Stdout:      stdout,
+		Stderr:      stderr,
+		Env:         env,
+		ExtraFiles:  extraFiles,
+		SysProcAttr: spa,
 	}
 
-	if err := cmd.Start(); err != nil {
-		channel.Close()
-		return nil, nil, err
+	if err = p.Cmd.Start(); err != nil {
+		return
 	}
 
 	// Reap the child when the process dies.
 	sc := make(chan os.Signal, 1)
 	signal.Notify(sc, syscall.SIGCHLD)
 	go func() {
-		<- sc
-		cmd.Wait()
+		<-sc
+		p.Cmd.Wait()
 		signal.Stop(sc)
+		os.RemoveAll(p.Tempdir)
+		p.Done <- true
+		close(p.Done) // prevent any more blocking
 	}()
 
-	return channel, cmd, nil
+	// TODO(kwalsh) put channel into p, remove the struct in linux_host.go
+
+	return
+}
+
+// ExitStatus returns an exit code for the process.
+func (p *HostedProcess) ExitStatus() (int, error) {
+	s := p.Cmd.ProcessState
+	if s == nil {
+		return -1, fmt.Errorf("Child has not exited")
+	}
+	if code, ok := (*s).Sys().(syscall.WaitStatus); ok {
+		return int(code), nil
+	}
+	return -1, fmt.Errorf("Couldn't get exit status\n")
```

Comment:
nit: it's probably more idiomatic Go to return 0 here, since a caller should check err before looking at the return value